// express-authenticators/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuth = exports.OAuth2 = exports.ZaloAuthenticator = exports.LineAuthenticator = exports.TwitterAuthenticator = exports.TumblrAuthenticator = exports.PinterestAuthenticator = exports.LinkedInAuthenticator = exports.InstagramAuthenticator = exports.GoogleAuthenticator = exports.GithubAuthenticator = exports.FoursquareAuthenticator = exports.FacebookAuthenticator = void 0;
const FacebookAuthenticator_1 = __importDefault(require("./vendors/FacebookAuthenticator"));
exports.FacebookAuthenticator = FacebookAuthenticator_1.default;
const FoursquareAuthenticator_1 = __importDefault(require("./vendors/FoursquareAuthenticator"));
exports.FoursquareAuthenticator = FoursquareAuthenticator_1.default;
const GithubAuthenticator_1 = __importDefault(require("./vendors/GithubAuthenticator"));
exports.GithubAuthenticator = GithubAuthenticator_1.default;
const GoogleAuthenticator_1 = __importDefault(require("./vendors/GoogleAuthenticator"));
exports.GoogleAuthenticator = GoogleAuthenticator_1.default;
const InstagramAuthenticator_1 = __importDefault(require("./vendors/InstagramAuthenticator"));
exports.InstagramAuthenticator = InstagramAuthenticator_1.default;
const LinkedInAuthenticator_1 = __importDefault(require("./vendors/LinkedInAuthenticator"));
exports.LinkedInAuthenticator = LinkedInAuthenticator_1.default;
const PinterestAuthenticator_1 = __importDefault(require("./vendors/PinterestAuthenticator"));
exports.PinterestAuthenticator = PinterestAuthenticator_1.default;
const TumblrAuthenticator_1 = __importDefault(require("./vendors/TumblrAuthenticator"));
exports.TumblrAuthenticator = TumblrAuthenticator_1.default;
const TwitterAuthenticator_1 = __importDefault(require("./vendors/TwitterAuthenticator"));
exports.TwitterAuthenticator = TwitterAuthenticator_1.default;
const LineAuthenticator_1 = __importDefault(require("./vendors/LineAuthenticator"));
exports.LineAuthenticator = LineAuthenticator_1.default;
const ZaloAuthenticator_1 = __importDefault(require("./vendors/ZaloAuthenticator"));
exports.ZaloAuthenticator = ZaloAuthenticator_1.default;
const OAuth_1 = __importDefault(require("./oauth/OAuth"));
exports.OAuth = OAuth_1.default;
const OAuth2_1 = __importDefault(require("./oauth2/OAuth2"));
exports.OAuth2 = OAuth2_1.default;
